hfst3/man/hfst-xfst.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.4.
.TH HFST-XFST "1" "February 2014" "HFST" "User Commands"
.SH NAME
hfst-xfst \- =Compile XFST scripts or execute XFST commands interactively
.SH SYNOPSIS
.B hfst-xfst
[\fIOPTIONS\fR...]
.SH DESCRIPTION
Compile XFST scripts or execute XFST commands interactively
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Only print fatal erros and requested output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.SS "Xfst-specific options:"
.TP
\fB\-e\fR, \fB\-\-execute\fR=\fICMD\fR
Execute command CMD on startup
.TP
\fB\-f\fR, \fB\-\-format\fR=\fIFMT\fR
Write result using FMT as backend format
.TP
\fB\-F\fR, \fB\-\-scriptfile\fR=\fIFILE\fR
Read commands from FILE, and quit
.TP
\fB\-l\fR, \fB\-\-startupfile\fR=\fIFILE\fR
Read commands from FILE on startup
.TP
\fB\-p\fR, \fB\-\-pipe\-mode\fR
Read commands from standard input (non\-interactive)
.TP
\fB\-r\fR, \fB\-\-no\-readline\fR
Do not use readline library for input
.TP
\fB\-k\fR, \fB\-\-output\-to\-console\fR
Output directly to console (Windows\-specific)
.PP
Option \fB\-\-execute\fR can be invoked many times.
If FMT is not given, OpenFst's tropical format will be used.
The possible values for FMT are { foma, openfst\-tropical, openfst\-log, sfst }.
Readline library, if enabled when configuring, is used for input by default.
Input files are always treated as UTF\-8.
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://sourceforge.net/tracker/?atid=1061990&group_id=224521&func=browse>
.PP
hfst\-xfst home page:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//HfstXfst2Fst>
.br
General help using HFST software:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//HfstHome>
.SH COPYRIGHT
Copyright \(co 2010 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
=================================
 NEWS for hfst-tools and library
=================================

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

Noteworthy changes in 3.3.9
---------------------------

* Bugs fixed and new features added to hfst-proc

* Xfst rules tentatively supported in hfst-regexp2fst

* hfst-fst2strings does not produce duplicate paths any more 
  when using foma as backend

* hfst-lexc implementation format defaults to openfst-tropical
  instead of foma

Noteworthy changes in 3.3.8
---------------------------

* Fixes 3.3.7

Noteworthy changes in 3.3.7
---------------------------

* Compatible with OpenFst version 1.3.1

* Identity symbols supported in hfst-regexp2fst

Noteworthy changes in 3.3.6
---------------------------

* Include directives of foma and SFST header files changed so
  that standard include path is used instead of source path

* Compilation warnings and some memory leaks fixed

Noteworthy changes in 3.3.5
---------------------------

* hfst-fst2txt.cc also writes to dot/graphviz format

* Epenthesis added for Xerox rules parser

* hfst-compose-intersect gives warnings on missing symbols

Noteworthy changes in 3.3.4
---------------------------

* Substitution functions optimized and new substitution functions added that
  perform multiple substitutions at the same time.

* The tool hfst-substitute now implements the option --in-order.

* Updated to OpenFst version 1.2.10.

* Conversions between transducer formats are more efficient.

Noteworthy changes in 3.3.3
---------------------------

* HfstTransitionGraph and optimized lookup transducer code optimized

* Foma transducer name is not copied to the net but is left empty

* Small fixes to twolc

Noteworthy changes in 3.3.2
---------------------------

* Forward compatibility with OpenFst 1.2.8 is achieved.

* PKG_ macros removed from end user builds since they do not work nicely on
  all platforms.

Noteworthy changes in 3.3.1
---------------------------

* Handling of flag diacritics improved.

* More examples in fsmbook-examples included in tests.

* Extra files created during tests as well as generated autotools files 
  are now ignored by svn.

Noteworthy changes in 3.3.0
---------------------------

* HFST updated to work with the newest versions of backend libraries

* a new replace down function by Karttunen added

* HfstTransitionGraph and conversions between formats made less error-prone

* protected/public changes in API

* SFST and foma libraries are disabled from default build

Noteworthy changes in 3.2.0
---------------------------

* Numerous improvements to hfst-proc, including better uppercasing via glib,
  better compound handling, new analysis mode for raw analysis without case
  changes or special symbols etc.

* old slow lexc (not built by default)

* numerous scripting fixes for mac os x compatibility with foma

* constness changes in API

Noteworthy changes in 3.1.1
---------------------------

* Numerous bugfixes, in particular regarding a broken assumption in
  compose-intersect

Noteworthy changes in 3.1.0
---------------------------

* api for lookup and lookdown changed, previously memory didn't get initialized
  anywhere, causing a serious bug

Noteworthy changes in 3.0.3
---------------------------

* numerous internal improvements, eg. distcheck is now functional

* many bugfixes, especially wrt. INDENTITY_PAIR & OTHER SYMBOL, composition,
  final weights in ol-converted transducers

Noteworthy changes in 3.0.2
---------------------------

* some fixes to OSX compatibility

* no more license conflict problems wrt foma

* better memory use in compose-intersect

* new implementation of internal transducer

Noteworthy changes in 3.0.1
---------------------------

* this is a bugfix release

* conversion to hfst-optimized-lookup format was fundamentally broken in the
  original release; this should provide a fix and considerable speedup

* compose-intersect memory use is much improved

Noteworthy changes in 3.0
-------------------------

* all new API in object oriented form

* brand new foma backend for unweighted transducer handling

* the FST libraries used for automata handling are now dynamically linked to
  HFST

    * this requires end users to install needed backends themselves

* hfst-lexc and hfst-xfst based on foma's compiler

  * xfst is now also fully supported

  * the obsolete command line clients for HFST-based compilation of lexc and
    xfst scripts are hidden in hfst-lexc2fst and hfst-xfst2fst

* preliminary support for unknown character classes

* optimized-lookup backend supported in library

  * dropped special tools for optimised handling since optimised lookup is
    included in library: hfst-optimised-lookup -> hfst-lookup; 
    hfst-lookup-optimize -> hfst-fst2fst

* hfst-proc tool to emulate lt-proc, cg-proc

* Slightly improved string and text formats to support weights uniformly and
  parsing tabs and spaces as automata arcs: see `wiki page for text and string
  formats
  <https://kitwiki.csc.fi/twiki/bin/view/KitWiki/HfstStringAndTextFormats>`

* better error handling using exceptions

* The argument handling order of binary tools changed: stdin is first if not
  specified

* The automata are now packed in HFST automaton container format (akin to AVI
  for videos); this supports arbitrary metadata and underlying FST formats

* the symbol handling is more implicit; no more symbol table arguments for tools
  or hfst-symbols command

* converted to new autotools: silent rules and color tests

* Hfst's internal automata format for fallback on some unsupported operations
  and for conversions

Noteworthy changes in 2.4.1
---------------------------

* Fixed inclusion path in pkg-config file

* Fixed extraneous epsilon path in hfst-lexc transducers

Noteworthy changes in release 2.4
---------------------------------


hfst3/tools/src/parsers/hfst-xfst.cc
//! @file hfst-xfst2fst.cc
//!
//! @brief Implemetation of command line program for parsing xfst commands

//       This program is free software: you can redistribute it and/or modify
//       it under the terms of the GNU General Public License as published by
//       the Free Software Foundation, version 3 of the License.
//
//       This program is distributed in the hope that it will be useful,
//       but WITHOUT ANY WARRANTY; without even the implied warranty of
//       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//       GNU General Public License for more details.
//
//       You should have received a copy of the GNU General Public License
//       along with this program.  If not, see <http://www.gnu.org/licenses/>.

#if HAVE_CONFIG_H
#   include <config.h>
#endif

#include "XfstCompiler.h"

#ifdef WINDOWS
  #include <io.h>
  #include <windows.h>
#endif

#include "../hfst-string-conversions.h"

#ifdef HAVE_READLINE
  #include <readline/readline.h>
  #include <readline/history.h>
#endif

#include <getopt.h>
#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "hfst-tool-metadata.h"
#include "inc/globals-common.h"
#include "hfst-file-to-mem.h"

static hfst::ImplementationType output_format = hfst::UNSPECIFIED_TYPE;
static char* scriptfilename = NULL;
static char* startupfilename = NULL;
static std::vector<char*> execute_commands;
static bool pipemode = false;
static bool output_to_console = false;
#ifdef HAVE_READLINE
  static bool use_readline = true;
#else
  static bool use_readline = false;
#endif
static bool print_weight = false;

void
print_usage()
{
  // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp                                                                                                                                                               
  // Usage line                                                                                                                                                                                                                            
  fprintf(message_out, "Usage: %s [OPTIONS...]\n"
          "Compile XFST scripts or execute XFST commands interactively\n"
          "\n", program_name);
  
  print_common_program_options(message_out);
  fprintf(message_out, "\n");
  fprintf(message_out, "Xfst-specific options:\n");
  fprintf(message_out, 
          "  -e, --execute=CMD        Execute command CMD on startup\n" 
          "  -f, --format=FMT         Write result using FMT as backend format\n"
          "  -F, --scriptfile=FILE    Read commands from FILE, and quit\n"
          "  -l, --startupfile=FILE   Read commands from FILE on startup\n"
          "  -p, --pipe-mode          Read commands from standard input (non-interactive)\n"
          "  -r, --no-readline        Do not use readline library for input\n"
          "  -w, --print-weight       Print weights for each operation\n"
          "  -k, --output-to-console  Output directly to console (Windows-specific)\n"
          "\n"
          "Option --execute can be invoked many times.\n"
          "If FMT is not given, OpenFst's tropical format will be used.\n"
          "The possible values for FMT are { foma, openfst-tropical, openfst-log, sfst }.\n"
          "Readline library, if enabled when configuring, is used for input by default.\n"
          "Input files are always treated as UTF-8.\n");
  fprintf(message_out, "\n");
  print_report_bugs();
  fprintf(message_out, "\n");
  print_more_info();
}

int
parse_options(int argc, char** argv)
{
  extend_options_getenv(&argc, &argv);
  // use of this function requires options are settable on global scope                                                                                                                                                                    
  while (true)
    {
        static const struct option long_options[] =
          {
            HFST_GETOPT_COMMON_LONG,
            // add tool-specific options here
            {"format", required_argument, 0, 'f'},
            {"scriptfile", required_argument, 0, 'F'},
            {"execute", required_argument, 0, 'e'},
            {"startupfile", required_argument, 0, 'l'},
            {"pipe-mode", no_argument, 0, 'p'},
            {"no-readline", no_argument, 0, 'r'},
            {"print-weight", no_argument, 0, 'w'},
            {"output-to-console", no_argument, 0, 'k'},
            {0,0,0,0}
          };
        int option_index = 0;
        // add tool-specific options here
        char c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT "f:F:e:l:prwk",
                             long_options, &option_index);
        if (-1 == c)
          {
            break;
          }

        switch (c)
          {
            //copied from "inc/getopt-cases-common.h"
          case 'd':
            debug = true;
            break;
          case 'h':
            print_usage();
            return EXIT_SUCCESS;
            break;
          case 'V':
            print_version();
            return EXIT_SUCCESS;
            break;
          case 'v':
            verbose = true;
            silent = false;
            break;
          case 'q':
          case 's':
            verbose = false;
            silent = true;
            break;
          case 'f':
            output_format = hfst_parse_format_name(optarg);
            break;
          case 'F':
            scriptfilename = hfst_strdup(optarg);
            break;
          case 'e':
            execute_commands.push_back(hfst_strdup(optarg));
            break;
          case 'l':
            startupfilename = hfst_strdup(optarg);
            break;
          case 'p':
            pipemode = true;
            use_readline = false;
            break;
          case 'r':
            use_readline = false;
            break;
          case 'w':
            print_weight = true;
            break;
          case 'k':
            output_to_console = true;
            break;
#include "inc/getopt-cases-error.h"
          }
    }

  if (output_format == hfst::UNSPECIFIED_TYPE)
    {
      output_format = hfst::TROPICAL_OPENFST_TYPE;
      verbose_printf("Using default output format OpenFst "
                     "with tropical weight class\n");
    }

  return EXIT_CONTINUE;
}

// Parse file 'filename' using compiler 'comp'.
// Filename "<stdin>" uses stdin for reading.
int parse_file(const char* filename, hfst::xfst::XfstCompiler &comp)
{
  char* line = hfst_file_to_mem(filename);
  if (NULL == line) 
    {
      error(EXIT_FAILURE, 0, "error when reading file %s\n", filename);
      return EXIT_FAILURE;
    }
  if (0 != comp.parse_line(line))
    {
      error(EXIT_FAILURE, 0, "error when parsing file %s\n", filename);
      return EXIT_FAILURE;
    }
  free(line);
  return 0;
}

void insert_zeroes(char * array, unsigned int number)
{
  for (unsigned int i=0; i<number; i++)
    {
      array[i] = '\0'; 
    }
}

static bool expression_continues(std::string & expr)
{
  size_t s = expr.size();
  if (s > 0 && expr.at(s-1) == '\\')
    {
      expr.at(s-1) = '\n';
      return true;
    }
  if (s > 1 && expr.at(s-1) == '\r' && expr.at(s-2) == '\\')
    {
      expr.at(s-1) = '\n';
      expr.at(s-2) = '\r';
      return true;
    }
  return false;
}


#ifdef HAVE_READLINE
#include "cmd.h"
#include "abbrcmd.h"

/* Variable to pass the position of rl completion to our completer */
static int smatch;

// By Mans Hulden
char *my_generator(const char *text, int state) {
  static int list_index, list_index2, len, nummatches;
  char *name;
  text = rl_line_buffer;
  if (!state) {
    list_index = 0;
    list_index2 = 0;
    nummatches = 0;
    len = strlen(text);
  }

  while ((name = cmd[list_index])) {
    list_index++;

    if (strncmp (name, text, len) == 0) {
      nummatches++;      
      return(strdup(name+smatch));
    }
  }

  if (rl_point > 0) {
    while ((name = abbrcmd[list_index2])) {
      list_index2++;

      if (strncmp (name, text, len) == 0)
        return(strdup(name+smatch));
    }
  }

  /* If no names matched, then return NULL. */
  return ((char *)NULL);
}


// By Mans Hulden
static char **my_completion(const char *text, int start, int end) {
  char **matches;

  matches = (char **)NULL;
  smatch = start;
  matches = rl_completion_matches ((char*)text, &my_generator);

  return (matches);
}
#endif // HAVE_READLINE

int main(int argc, char** argv)
{

  hfst_set_program_name(argv[0], "0.1", "HfstXfst2Fst");
  int retval = parse_options(argc, argv);
  if (retval != EXIT_CONTINUE)
    {
      return retval;
    }

  switch (output_format)
    {
    case hfst::SFST_TYPE:
      verbose_printf("Using SFST as output handler\n");
      break;
    case hfst::TROPICAL_OPENFST_TYPE:
      verbose_printf("Using OpenFst's tropical weights as output\n");
      break;
    case hfst::LOG_OPENFST_TYPE:
      verbose_printf("Using OpenFst's log weight output\n");
      break;
    case hfst::FOMA_TYPE:
      verbose_printf("Using foma as output handler\n");
      break;
    case hfst::HFST_OL_TYPE:
      verbose_printf("Using optimized lookup output\n");
      break;
    case hfst::HFST_OLW_TYPE:
      verbose_printf("Using optimized lookup weighted output\n");
      break;
    default:
      error(EXIT_FAILURE, 0, "Unknown format cannot be used as output\n");
      return EXIT_FAILURE;
    }
  
  if (pipemode && (scriptfilename != NULL))
    {
      error(EXIT_FAILURE, 0 , "--pipe-mode and --scriptfile cannot be used simultaneously\n");
      return EXIT_FAILURE;
    }

  if ((startupfilename != NULL) && (scriptfilename != NULL))
    {
      error(EXIT_FAILURE, 0, "--startupfile and --scriptfile cannot be used simultaneously\n");
      return EXIT_FAILURE;
    }


  // Create XfstCompiler
  hfst::xfst::XfstCompiler comp(output_format);
#ifdef HAVE_READLINE
  comp.setReadline(use_readline);
#else
  comp.setReadline(false);
#endif
  comp.setVerbosity(!silent);
  
  if (print_weight)
    {
      comp.setPromptVerbosity(false);
      comp.set("print-weight", "ON");
      comp.setPromptVerbosity(true);
    }

  if (output_to_console)
    comp.setOutputToConsole(true);

  // If needed, execute scripts given in command line
  for (std::vector<char*>::const_iterator cmd = execute_commands.begin();
       cmd != execute_commands.end(); cmd++)
    {
      verbose_printf("Executing xfst command '%s' given on command line\n", *cmd);
      if (0 != comp.parse_line(*cmd))
        {
          error(EXIT_FAILURE, 0, "command '%s' could not be parsed\n", *cmd);
          return EXIT_FAILURE;
        }
    }
  // If needed, execute script in startup file
  if (startupfilename != NULL)
    {
      verbose_printf("Executing startup file '%s'...\n", startupfilename);
      if (parse_file(startupfilename, comp) == EXIT_FAILURE)
        {
          return EXIT_FAILURE;
        }
    }

  if (pipemode) 
    {
      verbose_printf("Reading from standard input...\n");
      comp.setReadInteractiveTextFromStdin(false);
      if (parse_file("<stdin>", comp) == EXIT_FAILURE)
        //if (0 != comp.parse(stdin)) segfaults with scriptfiles..
        {
          return EXIT_FAILURE;
        }
    }
  else if (scriptfilename != NULL)
    {
      verbose_printf("Reading from script file '%s'\n", scriptfilename);
      if (parse_file(scriptfilename, comp) == EXIT_FAILURE)
        {
          return EXIT_FAILURE;
        }
    }
  // Use interactive mode
  else if (! use_readline)
    {
      verbose_printf("Starting interactive mode...\n");
      comp.setPromptVerbosity(!silent);
      comp.setReadInteractiveTextFromStdin(true);
      if (!silent)
        comp.prompt();
      // support for backspace

      std::string expression = "";
      //unsigned int lines = 0;

#ifdef WINDOWS
      SetConsoleCP(65001);
      const HANDLE stdIn = GetStdHandle(STD_INPUT_HANDLE);
      WCHAR buffer[0x1000];
      DWORD numRead = 0;
      while (ReadConsoleW(stdIn, buffer, sizeof buffer, &numRead, NULL))
        {
          std::wstring wstr(buffer, numRead);
          std::string linestr = hfst::wide_string_to_string(wstr);
          expression += linestr;
#else
      const unsigned int MAX_LINE_LENGTH = 1024;
      char line [MAX_LINE_LENGTH];
      insert_zeroes(line, MAX_LINE_LENGTH);
      while (cin.getline(line, MAX_LINE_LENGTH))
        {
          std::string linestr(line);
          expression += linestr;
#endif
          if (expression_continues(expression))
            {
              if (!silent)
                comp.prompt();
              continue;
            }

          if (0 != comp.parse_line((expression + "\n").c_str()))
            {
              fprintf(stderr, "expression '%s' could not be parsed\n", expression.c_str());
            }

          expression = "";
        }
    }
  else
    {
#ifdef HAVE_READLINE
      // support for backspace and Up/Down keys, needs readline library

      verbose_printf("Starting interactive mode...\n");
      comp.setPromptVerbosity(false); // prompts handled manually
      comp.setReadInteractiveTextFromStdin(true);
      char *buf = NULL;               // result from readline
      //rl_bind_key('\t',rl_abort);     // disable auto-complet
      std::string expression = "";    // the whole expression (possibly several lines)

      char* promptline = (!silent) ? comp.get_prompt() : strdup("");

      rl_attempted_completion_function = my_completion;

      while((buf = readline(promptline)) != NULL)
        {
          std::string bufstr(buf);
          expression += bufstr;

          if (expression.size() > 0 && expression.at(expression.size()-1) == '\\')
            {
              expression.at(expression.size()-1) = '\n';
              free(promptline);
              promptline = (!silent) ? comp.get_prompt() : strdup("");
              continue;
            }

          if (buf[0] != 0) {
            add_history(expression.c_str()); }

          if (0 != comp.parse_line((expression + "\n").c_str()))
            {
              fprintf(stderr, "expression '%s' could not be parsed\n", expression.c_str());
            }
          
          expression = "";
          free(promptline);
          promptline = (!silent) ? comp.get_prompt() : strdup("");
        }
      free(buf);
      free(promptline);
#else
      fprintf(stderr, "ERROR: missing readline library\n");
      return EXIT_FAILURE;
#endif
    }

  return EXIT_SUCCESS;
}

